addons/assigngear/script_component.hpp

```diff
@@ -40,6 +40,50 @@
 
 /* assignGear specific macros */
 #define GETGEAR(var) [_cfg >> var] call CFUNC(getCfgEntryFromPath)
+
+/* assignGear _loadout indexes */
+// Appearance
+#define IDX_DISPLAY_NAME 0
+#define IDX_FACES 1
+#define IDX_INSIGNIAS 2
+// Equipment
+#define IDX_HEADGEAR 3
+#define IDX_GOGGLES 4
+#define IDX_HMD 5
+// Containers - These must come before Items.
+#define IDX_UNIFORM 6
+#define IDX_VEST 7
+#define IDX_BACKPACK 8
+// Weapons & Weapon Magazine - These must come before magazines. 
+// Otherwise the weapon specific magazines will not work. E.g. PrimaryMagazine
+// Primary
+#define IDX_PRIMARY_WEAPON 9
+// Must come after PrimaryWeapon
+#define IDX_PRIMARY_MAGAZINE 10 
+#define IDX_PRIMARY_SCOPE 11
+#define IDX_PRIMARY_BIPOD 12
+#define IDX_PRIMARY_ATTACHMENT 13
+#define IDX_PRIMARY_SILENCER 14
+// Secondary
+#define IDX_SECONDARY_WEAPON 15
+// Must come after SecondaryWeapon
+#define IDX_SECONDARY_MAGAZINE 16 
+#define IDX_SECONDARY_ATTACHMENT 17
+// Sidearm
+#define IDX_SIDEARM_WEAPON 18
+// Must come after SidearmWeapon
+#define IDX_SIDEARM_MAGAZINE 19 
+#define IDX_SIDEARM_ATTACHMENT 20
+// Items & Magazines
+// Items go first to guarantee specific mission items are always included regardless of mission maker. E.g. Demo charges.
+#define IDX_ITEMS 21
+#define IDX_BACKPACK_ITEMS 22
+#define IDX_MAGAZINES 24
```

Comment:
Doesn't this break backwards compatibility for custom loadouts?

Reply:
Why would it? The lookups are all done by name?

Reply:
Wouldn't it cause units without primary/secondary/sidearmMagazine set to spawn without a magazine loaded?

Reply:
Yes, but a unit can reload and still have the same number of mags as before, the mag that was "loaded" into the weapon previously was being pulled from the unit's inventory and putting them a mag down effectively.

Reply:
Having to manually reload magazines after spawning on older missions/loadouts is going to become frustrating. I'd move `magazines` back before the `xMagazine` part, and do something like
```sqf
{
  _unit removePrimaryWeaponItem _x;
  _unit addItem _x; //TODO: Replace with magazines additem logic
} forEach primaryWeaponMagazine _unit;
{
  _unit addPrimaryWeaponItem _x;
} forEach _x;
```

This should preserve full backwards compatibility, albeit ever so slightly less performant.

Reply:
The problem with that is that you lose a magazine, which is the problem I'm trying to fix in this PR, the alternative would be to add a config flag to the top of the loadout file to enable or disable automatic loading from the player inventory.

> Having to manually reload magazines after spawning on older missions/loadouts is going to become frustrating.

I don't think the user impact will be massive, and the fix is very simple but if you think it's going to be too much we can look at other options.

Reply:
This is a breaking change. This issue must be resolved.

Reply:
> This is a breaking change. This issue must be resolved.

What's the preference to prevent the change from being breaking?

Reply:
Hmmmm, I think we can do this with some extra logic around the magazines, but it would require inspecting the weapon, seeing if it had a magazine and then adding that magazine to the inventory as part of adding the `primaryWeapon` logic. Is that acceptable?

Reply:
So we can do this with some logic applied to `assignGear`, however, it's not possible to guarantee that the magazine we remove can be added back to the unit. This occurs because we don't know where magazine being pulled in orginally came from, so if we put the magazines into the wrong container they may cause other magazines not fit in any container.

For example, suppose we have a primary weapon with a magazine (size 3) that was pulled from the player's vest and a UGL cartridge (size 1) that was pulled from the player's backpack. There is also a secondary weapons with a magazine (size 2) pulled from the player's vest.

If we process the primary weapon first, we pull out the magazines from the weapon and try and put both into the user's vest first, they both fit (3 + 1 = 4, 1 space left). When we process the secondary weapon, we no longer have space in the vest (2 > 1) and the space in the backpack is not enough to accept the magazine either (The UGL cartridge only guarantees 1 space is open in the backpack).

It's a complex problem to solve to avoid player's having to hit reload. Not sure if it's possible.

Reply:
Did you try my script snippet? I'm pretty sure it should work.

Reply:
It works, but can fail to add the magazine back to the unit, particularly for AT rounds. I'll push the code and if you're happy we can go forward with it.

Reply:
```sqf
{
    if !(_x isEqualTo '') then {
        _loadedMagazines = handgunMagazine _unit;
        if (count _loadedMagazines > 0) then {
            {
                _unit removeHandgunItem _x;
                switch true do {
                    case (_unit canAddItemToVest _x): {_unit addItemToVest _x;};
                    case (_unit canAddItemToUniform _x): {_unit addItemToUniform _x;};
                    default {_unit addItemToBackpack _x;};
                };
            } forEach _loadedMagazines;
        };
        {
            _unit addHandgunItem _x;
        } forEach _x;
    };
} forEach _x;
```

Reply:
> ```sqf
> {
>     if !(_x isEqualTo '') then {
>         _loadedMagazines = handgunMagazine _unit;
>         if (count _loadedMagazines > 0) then {
>             {
>                 _unit removeHandgunItem _x;
>                 switch true do {
>                     case (_unit canAddItemToVest _x): {_unit addItemToVest _x;};
>                     case (_unit canAddItemToUniform _x): {_unit addItemToUniform _x;};
>                     default {_unit addItemToBackpack _x;};
>                 };
>             } forEach _loadedMagazines;
>         };
>         {
>             _unit addHandgunItem _x;
>         } forEach _x;
>     };
> } forEach _x;
> ```

`count` is redundant as iterating over [] means it won't do anything. I don't think there is a usecase for checking if `_x == ''`, as it doesn't do anything different from `arr[] = {};`

Reply:
Ah yep, my bad.

Reply:
I think I've fixed it.

Reply:
I haven't looked into the intricacies yet but to throw another suggestion at it perhaps an alternative approach is that if `primaryMagazine` is not defined then move the first compatible magazine in the various magazine arrays and then cache it into `primaryMagazine` and removing it from the source magazine array. Although this has edge cases such as what if the randomized weapon isn't compatible.

Reply:
That's actually smart, we can just load the magazine at a later point unless it's explicitly set to being empty?

Reply:
> Although this has edge cases such as what if the randomized weapon isn't compatible.

This gets handled by the weapon checks. https://github.com/jdcaperon/TMF/blob/jack-loadout-add-weapon-magazine/addons/assigngear/functions/fnc_testGear.sqf#L226-L229

Reply:
Could just call https://community.bistudio.com/wiki/reload I guess 😂

Reply:
Same code as previous: 
https://github.com/jdcaperon/TMF/blob/jack-loadout-add-weapon-magazine-alt/addons/assigngear/functions/fnc_assignGear.sqf#L110-L112

We just force a reload of the unit after loadout is applied:
https://github.com/jdcaperon/TMF/blob/jack-loadout-add-weapon-magazine-alt/addons/assigngear/functions/fnc_assignGear.sqf#L166-L167